<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>

Андрей, доброго дня! рад приветствовать тебя вновь) <br />
(А дальше - шаблонная часть, как ты заметил по другим спринтам)<br>
<br />
Меня зовут Николай Шавлюго. <br />И на этом этапе твоего движения к новой профессии от меня будут комментарии по написанному коду. <br />Чтобы меньше "мусорить эфир" и чтобы максимально наглядно отделяться от основного кода, есть предложение условиться в некоторых обозначениях:<br /> <br />
<u><b>ТАКОЙ ШРИФТ</b></u> - всегда начало комментария <br />
<font color='green'>такой шрифт</font> - комментарии о том, что всё ОК <br/>
<font color='orange'>такой шрифт</font> - комментарии о том, что всё ОК по результату,<br> однако есть на что обратить внимание в плане применения техник, или есть способы сделать более короткий или быстрый код <br/>
<font color='red'>такой шрифт</font> - комментарии о том, что есть критичный момент, влияющий на бизнес-результат проекта.<br/>
</font><br /><br />
<font color='green'>Моей целью является не "уличить" в не знании, а просто высказать сверху твоих знаний - свой опыт, что бы ты мог использовать его для своего дальнейшего успеха) И очень здорово будет, если тебе удастся задавать вопросы, да и вообще - всячески доставать меня, если я по каким-то причинам не приму проект:) При этом, из своего опыта скажу, хорошо и важно, когда переписка ревьюера и студента - сохраняется на следующие проекты и даже на будущую практическую деятельность.<br>
<BR> В ПУТЬ!<br>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Загрузка-и-проверка-данных" data-toc-modified-id="Загрузка-и-проверка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка и проверка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Импорт библиотек

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

## Загрузка и проверка данных

In [2]:
data = pd.read_csv('/datasets/insurance.csv')
def line():
    print()
    print('_________________________________')

In [3]:
data.head(5)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [24]:
display(data.info)
line()
display('Количество пропусков', data.isnull().sum())
line()
display('Количество дубликатов ', data.duplicated().sum())
line()
display(data.describe())
line()


<bound method DataFrame.info of       Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0       1     41.0   49600.0            1                  0
1       0     46.0   38000.0            1                  1
2       0     29.0   21000.0            0                  0
3       0     21.0   41700.0            2                  0
4       1     28.0   26100.0            0                  0
...   ...      ...       ...          ...                ...
4995    0     28.0   35700.0            2                  0
4996    0     34.0   52400.0            1                  0
4997    0     20.0   33900.0            2                  0
4998    1     22.0   32700.0            3                  0
4999    1     28.0   40600.0            1                  0

[5000 rows x 5 columns]>


_________________________________


'Количество пропусков'

Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64


_________________________________


'Количество дубликатов '

153


_________________________________


Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000


_________________________________


Страховые выплаты
0    4436
1     423
2     115
3      18
4       7
5       1
dtype: int64

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
Данные на базе) ОК, хорошо, когда есть аналитическая культура - обязательно делать первичный анализ данных, не смотря на проект)

### Вывод

- Имеем датасет из 5000 строк и 5 столбцов
- Пропуски отсутствуют
- Дубликаты 153 штуки. Дубликаты решил пока не удалять. считаю что они не повлияют на результат иследования 
- Целевой признак: Страховые выплаты
- Признаки: Пол, Возраст, Зарплата, Члены семьи
- Похоже, что данные предварительно обработаны

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** Не изменится

**Обоснование:** 
- Обозначим новую матрицу признаков X1 как произведение старой X на матрицу P(обратимую):

$$
X1 = X*P
$$ 

- Подставим новое значение X1 в формулу w1:

$$
w1 = ((XP)^T XP)^{-1} (XP)^T y
$$

- Раскроем первое произведение (XP)^T: 

$$
w1 = (X^TP^T XP)^{-1} X^TP^T y
$$

- Раскроем скобки

$$
w1 = P^{-1}(X^TX)^{-1}(P^T)^{-1} P^T X^T y
$$
            
- Поусловию $ Р $ обратима поэтому произведение $ (P^T)^{-1}P^T $ равно единичной матрице $E$

$$
w1 = P^{-1}(X^TX)^{-1} EX^T y =  P^{-1}(X^TX)^{-1} X^T y 
$$

- Справа получилась формула ля $w$

$$
w1 = P^{-1}(X^TX)^{-1} EX^T y =  P^{-1}w
$$

- В формулу для предсказаний подставим $X1$ и $w1$ что бы вычислить предсказания модели $a1$

$$
a1 = X1w1=XPP^{-1}w
$$

- По скольку по условию $ Р $ обратима поэтому произведение $PP^{-1}$ равно единичной матрице $E$

$$
a1 = X1w1=XPP^{-1}w = XEw=Xw=a
$$


Было доказано, что предсказания $а1$ для матрицы признаков, умноженных на обратимую матрицу $ Р $ равны предсказаниям $а$

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
в точку!

## Алгоритм преобразования

**Алгоритм**

Для того чтобы закодировать данные пользователя, можно создать рандомно генирирующуюся матрицу которая имеет обратную себе матрицу. Таким образом домножая матрицу признаков на рандомную матрицу матрицу мы получим матрицу с закодированными значениями а для того чтобы получить обратно данные мы должны будем умножить закодированные данные на обратную рандомную матрицу

**Обоснование**

$$
a1 = X1w1=XPP^{-1}w = XEw=Xw=a
$$


In [5]:
#data.drop_duplicates(inplace = True)

In [6]:
# разделение признаков
target = data.loc[:, 'Страховые выплаты']
features = data.drop(['Страховые выплаты'], axis=1)

In [7]:
# разделение выборок
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.2, random_state=12345)

In [8]:
# преобразование выборок в матрицы
matrix_features_train = features_train.values
matrix_features_test = features_test.values
# проверка размерности матрицы
print(matrix_features_train.shape)
print(matrix_features_test.shape)

(4000, 4)
(1000, 4)


In [9]:
# создание рандомной матрицы в соответствии с кол-вом признаков исходной матрицы
random_matrix = np.random.randn(4, 4)
# умножение исходных матриц на рандомную
encoded_matrix_features_train = matrix_features_train.dot(random_matrix)
encoded_matrix_features_test = matrix_features_test.dot(random_matrix)

In [10]:
# проверка
decoded_matrix_features_train = encoded_matrix_features_train.dot(np.linalg.inv(random_matrix))
features_train_check = pd.DataFrame(decoded_matrix_features_train, columns=features_train.columns)
features_train_check = np.round(features_train_check)
features_train_check.head(5), features_train.head(5)

(   Пол  Возраст  Зарплата  Члены семьи
 0 -0.0     20.0   39800.0          2.0
 1  1.0     40.0   34200.0         -0.0
 2  1.0     45.0   50800.0          1.0
 3 -0.0     28.0   39100.0          2.0
 4  0.0     28.0   39000.0          1.0,
       Пол  Возраст  Зарплата  Члены семьи
 317     0     20.0   39800.0            2
 4143    1     40.0   34200.0            0
 4252    1     45.0   50800.0            1
 710     0     28.0   39100.0            2
 148     0     28.0   39000.0            1)

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
ок, это важно - проверять на обратимость) Можно это дело в такую функцию одеть:

<div class="alert alert-info"> <b>Комментарий студента:</b> Большое спасибо за подсказку в будущем обязательно буду пользоваться таким методом </div>

Данные идентичны. только с индексами какая-то беда случилась.

## Проверка алгоритма

In [11]:
model = LinearRegression()
model.fit(matrix_features_train, target_train)
predicted_norm = model.predict(features_test)
norm_r2 = model.score(features_test, target_test)

In [12]:
model.fit(encoded_matrix_features_train, target_train)
predicted_encoded = model.predict(encoded_matrix_features_test)
encoded_r2 = model.score(encoded_matrix_features_test, target_test)

In [13]:
print('R2 score of not encoded matrix: ', norm_r2)
print('R2 score of encoded matrix: ', encoded_r2)

R2 score of not encoded matrix:  0.4117683956770434
R2 score of encoded matrix:  0.4117683956769309


<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА</u></b>
</font>
<font color='green'><br>
ОК, Андрей, и с линейной алгеброй ты справился)<br>
Всё, на мой взгляд, по делу, со смыслом. Доказательство в п.2 - понятно)<br>
Да и весь проект - на уровне.<br>
И важно, что не забыл про проверку на обратимость! ОК<br>
    
В общем-то, вопросов у меня и нет особых)<br>
Возможно, что у тебя остались/появились?

<div class="alert alert-info"> <b>Комментарий студента:</b> Добрый день! Хотел бы задать прару вопросов.<br> 
1. При востановлении матрицы  индексы меняются. В текущей задаче это не вожно, но в принцыпе в индекс могли быть заложены какие то данные по которым возможна идентификация субъекта при работе с персональными данными это очень важно. Есть ли способ избежать изменения индексов или способ их востановить?<br>
2. Вопрос про дубликаты. Подскажи, пожалуйста, какие именно дубликаты считает этот код data.duplicated().sum())? ведь например если посчитать дубликаты по каждому столбцу их будет намного больше. И почему дубликаты так страшны в иследованиях. Может быть есть какая-то толковая статья (на русском языке) на этот счет?  </div>

<div style="background: #cceeaa; padding: 5px; border: 1px solid green; border-radius: 5px;">
    <font color='green'> <b><u>КОММЕНТАРИЙ РЕВЬЮЕРА 2</u></b>
</font>
<font color='green'><br>
Андрей,  доброго дня!<br>
    
1) Тут надо от данных будет смотреть. Ведь всегда можно одним из столбцов сделать id клиента. (но не включать его в модель). И по этому столбцу отслеживать качество данных.<br>    
2) Дубликаты: тут надо исходить из определения дубликатов. Строка исходных данных - это описание одного клиента (например). И нам надо удостовериться, что нет одинаковых строк. Условно Иванова учтём один раз, а не двадцать. Со статьями - тут надо к наставнику - у них формируеътся список при подготовке к лекции, который они раздают студентам.<br>
    
Ещё несколько идей в целом по теме: если рассматривать передачу данных третьим лицам, то к преимуществам данного подхода можно отнести, в первую очередь, ощутимую экономию ресурсов по сравнению с некоторыми известными алгоритмами шифрования, в особенности это касается ассиметричных алгоритмов с достаточно длинными ключами (AES, RSA), которые требуют много процессорного времени.<br><br>
    
К главному недостатку подхода можно отнести увеличение сложности проверки на вырожденность преобразования с ростом количества признаков в выборке. В рассмотренном случае необходимо подобрать подходящую матрицу размера  4×4 . Однако, если признаки будут исчисляться тысячами, что в теории всегда надо положить, то расчет соответствующих определителей может быть крайне ресурсозатратным. Но и эту проблему можно будет решить)<br><br>
    
Если есть время и желание, то можно воту этот маткриал просмотреть:<br>
Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf, https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography
Подробнее про аналогичный алгоритм: https://www.cs.cmu.edu/~rjhall/JOS_revised_May_31a.pdf<br><br>
<font color='green'><font color='green'>
Спасибо за проект и за диалог)<br>
Успешного дальнейшего обучения!

### Вывод

При проверке качества моделей Линейной регрессии с помощью метрики R2 на данных показатели метрики совпали. Из чего можно сделать вывод, что алгоритм работает. Данный алгоритм можно использовать для шифрования данных.


Значение метрики при удалении дубликатов
- R2 score of not encoded matrix:  0.41605492161510915
- R2 score of encoded matrix:  0.4160549216152173
- На мой взгляд в данном случае удаление дубликатов не влияет на конечный результат

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования